fw/beef.cpp

```diff
@@ -20,20 +24,61 @@ uint16_t button_state;
 bool ignore_buttons;
 AbstractUsbHandler* usb_handler;
 button_pins buttons[] = CONFIG_ALL_HW_PIN;
+Command current_command;
+
+bool run_bootloader ATTR_NO_INIT;
+
+HidReport<config, INTERFACE_ID_Config, ENDPOINT_CONTROLEP> config_hid_report;
 
 ISR(TIMER1_COMPA_vect) {
   milliseconds++;
 }
 
-void debounce(DebounceState &debounce, uint16_t mask) {
-  button_state =
-    (button_state & ~mask) |
-    (debounce.debounce(button_state & mask));
+void reboot() {
+  USB_Detach();
+  wdt_enable(WDTO_250MS);
+  while (true);
+}
+
+void Application_Jump_Check() {
+  // Check if the reset source was from the watchdog
+  // and if we received a command/button combo to reset to bootloader
+  if ((MCUSR & (1 << WDRF)) && run_bootloader) {
+    MCUSR &= ~(1 << WDRF);
+    wdt_disable();
+
+    run_bootloader = false;
+    jump_to_bootloader();
+  }
+}
+
+void RebootToBootloader() {
+  reactive_led = false;
+  update_button_lighting(0);
+  FastLED.clear(true);
+
+  run_bootloader = true;
+  reboot();
 }
 
 void check_for_dfu() {
```

Comment:
Styling concern, I noticed several styles of function naming:

`lower_snake` -> `void check_for_dfu`
`Upper_Sanke` -> `void Application_Jump_Check`
`CamelCase` -> `void RebootToBootloader`

I have no preference as long as we're consistent.

Reply:
Yeah, it's an issue I've noticed when I'm using the LUFA demos and other projects as a reference. I've been trying to keep naming styles consistent with LUFA and other LUFA projects so that the code is easier to cross-reference against.

Looks like LUFA APIs are `Upper_Snake` while in most LUFA projects/demos application-specific functions are `CamelCase`. It looks like we're the outlier with `lower_snake`, but I'll change our functions to `lower_snake` to make it consistent.

Reply:
Actually, looks like LUFA has a mix of `Upper_Snake` and `CamelCase`. I guess, ideally, we'd be using `CamelCase` but it'd probably be a bit much to rename every function to that style.